runtime/drive_handler.go

```diff
@@ -0,0 +1,145 @@
+// Copyright 2018-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync/atomic"
+
+	firecracker "github.com/firecracker-microvm/firecracker-go-sdk"
+	models "github.com/firecracker-microvm/firecracker-go-sdk/client/models"
+
+	"github.com/firecracker-microvm/firecracker-containerd/internal"
+)
+
+const (
+	minFCDriveImageBytes = 512
+)
+
+var (
+	// ErrStubDrivesExhausted occurs when there are no more stub drives left to
+	// use. This can happen by calling PatchStubDrive greater than the stub drive
+	// list.
+	ErrStubDrivesExhausted = fmt.Errorf("There are no remaining stub drives to use")
+
+	// ErrDriveIDNil should never happen, but we safe guard against nil dereferencing
+	ErrDriveIDNil = fmt.Errorf("DriveID of current drive is nil")
+)
+
+// stubDriveHandler is used to manage stub drives.
+type stubDriveHandler struct {
+	RootPath       string
+	stubDriveIndex int64
+	drives         []models.Drive
+}
+
+func newStubDriveHandler(path string) stubDriveHandler {
+	return stubDriveHandler{
+		RootPath: path,
+	}
+}
+
+// StubDrivePaths will create stub drives and return the paths associated with
+// the stub drives.
+func (h *stubDriveHandler) StubDrivePaths(amount int) ([]string, error) {
+	paths := []string{}
+	for i := 0; i < amount; i++ {
+		driveID := fmt.Sprintf("stub%d", i)
+		path := filepath.Join(h.RootPath, driveID)
+
+		f, err := os.OpenFile(path, os.O_CREATE|os.O_EXCL|os.O_RDWR, 0600)
+		if err != nil {
+			return nil, err
+		}
+
+		if err := internal.GenerateStubContent(f, driveID); err != nil {
+			f.Close()
+			return nil, err
+		}
+
+		// Firecracker will not show any drives smaller than 512 bytes. In
+		// addition, the drive is read in chunks of 512 byes; if the drive size is
+		// not a multiple of 512 byes, then the remainder will not be visible to
+		// Firecracker.
+		if err := os.Truncate(path, minFCDriveImageBytes); err != nil {
```

Comment:
We need to verify somewhere that the stub content doesn't exceed 512 blocks, since any content beyond that gets truncated.  That can either happen inside StubDrivePaths or GenerateStubContent.